linear_operator/operators/block_tensor_linear_operator.py

```diff
@@ -0,0 +1,145 @@
+            for i in range(T):
+                for j in range(T):
+                    out_ij = self.linear_operators[i][0] @ rhs.linear_operators[0][j]
+                    for k in range(1, T):
+                        out_ij += self.linear_operators[i][k] @ rhs.linear_operators[k][j]
+                    output[i][j] = out_ij
+            return self.__class__(output)
+        elif isinstance(rhs, Tensor) and rhs.ndim == 2:
+            # Check both matrix dims divisible by T,
+            # reshape to (T, T, ), call block multiplication
+            if rhs.size(0) % T == 0 and rhs.size(1) % T == 0:
+                # A is block [N * T, M * T] and B is a general tensor/operator of shape [O, P].
+                # If O and P are both divisible by T,
+                # then interpret B as a [O//T * T, P//T * T] block matrix
+                O_T = rhs.size(0) // T
+                P_T = rhs.size(1) // T
+                rhs_blocks_raw = rhs.reshape(T, O_T, T, P_T)
+                rhs_blocks = rhs_blocks_raw.permute(0, 2, 1, 3)
+                rhs_op = BlockTensorLinearOperator.from_tensor(rhs_blocks, T)
+                return self._matmul(rhs_op)
+
+        # Failover implementation. Convert to dense and multiply matricies
+        A = self.to_dense()
+        B = rhs.to_dense()
+        res = A @ B
+        return res
+
+    def matmul(
+        self: Float[LinearOperator, "*batch M N"],
+        other: Union[Float[Tensor, "*batch2 N P"], Float[Tensor, "*batch2 N"], Float[LinearOperator, "*batch2 N P"]],
+    ) -> Union[Float[Tensor, "... M P"], Float[Tensor, "... M"], Float[LinearOperator, "... M P"]]:
+        return self._matmul(other)
+
+    def to_dense(self: Float[LinearOperator, "*batch M N"]) -> Float[Tensor, "*batch M N"]:
+        out = []
+        for i in range(self.num_tasks):
+            rows = []
+            for j in range(self.num_tasks):
+                rows.append(self.linear_operators[i][j].to_dense())
+            out.append(torch.concat(rows, axis=1))
+        return torch.concat(out, axis=0)
+
+    def _size(self) -> torch.Size:
+        sz = self.linear_operators[0][0].size()
+        return torch.Size([self.num_tasks * sz[0], self.num_tasks * sz[1]])
+
+    @property
+    def dtype(self) -> Optional[torch.dtype]:
+        return self.linear_operators[0][0].dtype
+
+    @property
+    def device(self) -> Optional[torch.device]:
+        return self.linear_operators[0][0].device
+
+    def representation(self) -> Tuple[torch.Tensor, ...]:
+        """
+        Returns the Tensors that are used to define the LinearOperator
+        """
+        representation = []
+        for op_row in self.linear_operators:
+            for op in op_row:
+                representation += tuple(op.representation())
+        return tuple(representation)
+
+    def _diag(self):
+        out = []
+        for i in range(self.num_tasks):
+            diagonal = self.linear_operators[i][i].diagonal()
```

Comment:
I assume we don't need to check the LO is square here? We can just leave these to the underlying LO to check?

Reply:
I think this can apply even if the underlying operators are non square. For example, PyTorch defines a general diagonal like this:
https://pytorch.org/docs/stable/generated/torch.diagonal.html

Reply:
I agree, though, that if we have non-square operators this implementation does not work.  As you mention, and non-square underlying operators should throw an error when _diagonal is called.

Reply:
Examinine there code, I think the is_square method will work on our new class. Assuming all the linear operators have the same shape the is_sqaure code just checks the first linear operator to see if it has square dimensions.
```
    @property
    def is_square(self) -> bool:
        return self.matrix_shape[0] == self.matrix_shape[1]
```